### PR TITLE
HPCC-26497 Cannot pass attributes to roxie queries on soapcalls

### PR DESCRIPTION
**Pull request by richardkchapman**
Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
Manually via test forms. This change should make it easier to add a checkbox to say "gather stats" when submitting a Roxie query via test forms.

### CODE REVIEW

**Review by afishbeck — APPROVED** [review]
@richardkchapman approved.

File: roxie/ccd/ccdprotocol.cpp [hunk 1]
```diff
@@ -1977,6 +1977,11 @@ class RoxieSocketWorker : public ProtocolQueryWorker
                                     {
                                         fixedreq->addPropTree(iter->query().queryName(), LINK(&iter->query()));
                                     }
+                                    Owned<IAttributeIterator> aiter = queryPT->getAttributes();
```
Comment (ghalliday) [inline roxie/ccd/ccdprotocol.cpp:1980]: cut and paste: Should be reqIter->query().getttributes()

**Review by ghalliday — CHANGES_REQUESTED** [review]

File: roxie/ccd/ccdprotocol.cpp (hunk 1 above)
Comment (richardkchapman) [reply roxie/ccd/ccdprotocol.cpp:1980]: Maybe - it's debatable whether attributes are common to every request in the array, or specified on each one. I can see arguments for both (or maybe even merge them).

You could also argue that for something like "statsToWorkunit" we should be add additional logic to make all the subqueries in the array write to the same workunit. But that would be something for a future enhancement.

Comment (richardkchapman) [reply roxie/ccd/ccdprotocol.cpp:1980]: @afishbeck Do you have an opinion?

**Comment by GordonSmith** [comment]
Testing this PR and I am seeing:

Request:
```
<?xml version="1.0" encoding="UTF-8"?>
<soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/" xmlns:SOAP-ENC="http://schemas.xmlsoap.org/soap/encoding/">
 <soap:Body>
  <roxieoverview1Request wuid="*" xmlns="urn:hpccsystems:ecl:roxieoverview1"/>
 </soap:Body>
</soap:Envelope>
```

Response:
```
<?xml version="1.0" encoding="UTF-8"?>
<soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/">
 <soap:Body>
  <roxieoverview1Response xmlns="urn:hpccsystems:ecl:roxieoverview1">
   <Results>
    <Result>
     <Exception>
      <Source>Roxie</Source>
      <Code>25</Code>
      <Message>SDS: IPropertyTree exception 
SDS Reply Error  : SDS: IPropertyTree exception 
IPropertyTree: Ambiguous xpath used
getProp: ambiguous xpath &quot;*&quot; in xpath '/WorkUnits/*'</Message>
     </Exception>
    </Result>
   </Results>
  </roxieoverview1Response>
 </soap:Body>
</soap:Envelope>
```

**Comment by GordonSmith** [comment]
Further - supplying a dummy WU worked (attached some metrics) once but didn't output any response, doing it a second time resulted in another error:
```
<?xml version="1.0" encoding="UTF-8"?>
<soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/">
 <soap:Body>
  <roxieoverview1Response xmlns="urn:hpccsystems:ecl:roxieoverview1">
   <Results>
    <Result>
     <Exception>
      <Source>Roxie</Source>
      <Code>0</Code>
      <Message>IPropertyTree: Ambiguous xpath used
getProp: ambiguous xpath &quot;Variable[@name='firstname']&quot;</Message>
     </Exception>
    </Result>
   </Results>
  </roxieoverview1Response>
 </soap:Body>
</soap:Envelope>
```

**Comment by GordonSmith** [comment]
Final note:  To pass an attr in via JSON the following looks like it might work:

```
{
 "roxieoverview1": {
   "@wuid": "*",
   "firstname": null,
   "lastname": null
 }
}
```

File: roxie/ccd/ccdprotocol.cpp (hunk 1 above)
Comment (afishbeck) [reply roxie/ccd/ccdprotocol.cpp:1980]: I like the way it is now.  The attributes can be set at the array level, and/or overridden at the individual request.

**Comment by afishbeck** [comment]
@GordonSmith your example JSON looks right to me.

**Comment by richardkchapman** [comment]
@GordonSmith I think I misled you when I said that passing wuid='*' would work.

What you need to pass is statsToWorkunit='1' - you can then pass in the wuid of the workunit you want the stats added to if you want. If you don't supply a workunit one will be created.

**Review by ghalliday — APPROVED** [review]
@richardkchapman please squash

**Pull request merged**